### ISSUE TITLE
Compatibility with Cloudflare Calls (TURN & STUN servers)

### ISSUE THREAD

**Issue by HyCore** [nextcloud/spreed#12394]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Feature request

**Which Nextcloud Version are you currently using:** (see administration page)
Latest (28.0.5)

**Is your feature request related to a problem? Please describe.**
Right now selfhosted TURN & STUN performance aren't great even with a top of the line server, 32 core & more, i often find myself glitched out in calls with my team. That's why when we saw a high performance low latency alternative like Cloudflare calls it would just make sense.

**Describe the solution you'd like**
Integration of Cloudflare Calls TURN service into Spreed.
It require a curl to their server like the following:
```bash
curl -X POST \
$TURN_SERVICE_API_TOKEN" \
	-H "Content-Type: application/json" -d '{"ttl": 86400}' \
$TURN_SERVICE_ID/credentials/generate
```
And would answer the credential for each user in Spreed (if they have different ip)

```json
{
  "iceServers": {
    "urls": [
      "stun:stun.cloudflare.com:3478",
      "turn:turn.cloudflare.com:3478?transport=udp",
      "turn:turn.cloudflare.com:3478?transport=tcp",
      "turns:turn.cloudflare.com:5349?transport=tcp"
    ],
    "username": "XXXXX",
    "credential": "YYYYY"
  }
}
```

and as a last step, if the creds are valid, it return the turn & stun connection to use & work with.

```python
const myPeerConnection = new RTCPeerConnection({
  iceServers: [
    {
      urls: "stun:stun.cloudflare.com:3478",
    },
    {
      urls: "turn:turn.cloudflare.com:3478",
      username: "REPLACE_WITH_USERNAME",
      credential: "REPLACE_WITH_CREDENTIAL",
    },
    {
      urls: "turns:turn.cloudflare.com:443?transport=tcp",
      username: "REPLACE_WITH_USERNAME",
      credential: "REPLACE_WITH_CREDENTIAL",
    },
    {
      urls: "turn:turn.cloudflare.com:3478?transport=tcp",
      username: "REPLACE_WITH_USERNAME",
      credential: "REPLACE_WITH_CREDENTIAL",
    },
  ],
});
```

I hope such feature would come available in a "near" future :D It would be so great.

Have a nice day ^^

**Comment by ther3zz** [comment]
I would absolutely love this as well

**Comment by dpurnam** [comment]
Bump.

Also, do I understand correctly that integrating this could potentially interest all those users who are unable to setup/manage an HPB instance for that perfect **Talk Call** experience?

Regardless, thank you very much for improving/maintaining Spreed over the years.

**Comment by nickvergessen** [comment]
> do I understand correctly that integrating this could potentially interest all those users who are unable to setup/manage an HPB instance for that perfect Talk Call experience?

No, it still requires the HPB to deduplicate video uploads and improve general performance of Talk

**Comment by wegylexy** [comment]
And not just Cloudflare. There are several TURN services that work in this way, providing a REST API like this to generate temporary credentials. So at the minimum, please take a URL, send a POST request per connection, parse the JSON and so on. In case the JSON structure is different, we could transform it via a reverse proxy until you support them.

**Comment by wegylexy** [comment]
I tried injecting code in `Config.php` and got `iceServers` on the clients (have to reload the page for a new set of credentials), but they time out. Is there something else I need to config?

**Comment by wegylexy** [comment]
Even better maybe to use [Cloudflare SFU](https://developers.cloudflare.com/realtime/calls-vs-sfus/) directly without clumsy TURN nor STUN.

**Comment by derStephan** [comment]
in #3301 @fancycode  wrote: 

> That's why Talk switched to the shared secret TURN configuration and hands out short-lived credentials to users that can not be used indefinitely.

This sounds exactly as the approach that Cloudflare describes oin how to get short-lived username/password from an TURN Key:
https://developers.cloudflare.com/realtime/turn/generate-credentials/

So will it suffice to put the API URL together with the TURN key in the talk setting and it will create username/passwords for each client automatically?

Regards
Stephan

**Comment by derStephan** [comment]
@fancycode just gave me the hint in the other ticket that the priciple is the same but it is not compatible at the moment. 

I suggest to write a PR to https://github.com/nextcloud/spreed/blob/main/lib/Config.php#L485 that checks if the Server URL is turn.cloudflare.com and calls the cloudflare API instead of the current approach. 

I know this is hacky but I would give it a try. Maybe this could serve as a starting point. What do you think?
